File: 3. Behavioral Design Patterns/Observer/Go/ObserverTemplate.go
package main

// Observer Pattern – Go Template

type Observer interface{ Update() }
type Subject interface{ Attach(Observer); Detach(Observer); Notify() }

type ConcreteSubject struct{ observers []Observer }
func (s *ConcreteSubject) Attach(o Observer){ s.observers = append(s.observers, o) }
func (s *ConcreteSubject) Detach(o Observer){ for i, obs := range s.observers { if obs == o { s.observers = append(s.observers[:i], s.observers[i+1:]...); break } } }
func (s *ConcreteSubject) Notify(){ for _, o := range s.observers { o.Update() } }
func (s *ConcreteSubject) DoSomething(){ /* change state */ s.Notify() }

type ConcreteObserver struct{}
func (ConcreteObserver) Update(){ /* react to change */ }

func main(){ s := &ConcreteSubject{}; s.Attach(ConcreteObserver{}); s.DoSomething() }

File: 3. Behavioral Design Patterns/Observer/Typescript/ObserverTemplate.ts
// Observer Pattern – TypeScript Template

interface Observer { update(): void }
interface Subject { attach(o: Observer): void; detach(o: Observer): void; notify(): void }

class ConcreteSubject implements Subject {
  private observers: Observer[] = []
  attach(o: Observer){ this.observers.push(o) }
  detach(o: Observer){ this.observers = this.observers.filter(obs => obs !== o) }
  notify(){ for(const o of this.observers) o.update() }
  doSomething(){ /* change state */ this.notify() }
}

class ConcreteObserver implements Observer { update(){ /* react to change */ } }

;(function main(){ const s = new ConcreteSubject(); s.attach(new ConcreteObserver()); s.doSomething() })()

export {}
